src/coreclr/vm/jitinterface.cpp

```diff
@@ -4919,6 +4919,13 @@ void CEEInfo::getCallInfo(
     pResult->stubLookup.lookupKind.needsRuntimeLookup = false;
 
     MethodDesc* pMD = (MethodDesc *)pResolvedToken->hMethod;
+    if (flags & CORINFO_CALLINFO_RUNTIMEASYNC_VARIANT)
+    {
+        _ASSERTE(!pMD->IsAsync2Method());
+        pMD = pMD->GetAsyncOtherVariant();
+        pResolvedToken->hMethod = (CORINFO_METHOD_HANDLE)pMD;
+    }
+
```

Comment:
Nice, that's much simpler than I was expecting.

`pResolvedToken` is in-only, so we should make a copy of it here and change that one instead. If necessary you can update it from the `callInfo` on the JIT side, but I'm somewhat worried we end up with a token whose fields are internally inconsistent.

Can you make sure we have tests for some of the hard cases? GVMs, interface calls and virtual class calls come to mind. I was expecting shared generics to require more work as well since other fields of the token are used below for those (see `ComputeRuntimeLookupForSharedGenericToken`). Can you double check why it works out? Is the method spec/type spec ok to reuse as-is from the token?

Reply:
Another way to ensure the resolved token consistency could be to pass the new flag not to the `eeGetCallInfo`, but to the `impResolveToken`.

Reply:
I've moved the MethodDesc shimming to the level of `impResolveToken`. That seems nicer as it allows `eeGetCallInfo` to stay unchanged.